### PR TITLE
Draft: Glucose forecasting example (BGI Open Build / AGI-26)

### PR DESCRIPTION
**Pull request by antonkulaga**
## Context — BGI Open Build / AGI-26

This is a **draft** pull request submitted as part of our [BGI Open Build (AGI-26 edition)](https://mailchi.mp/singularitynet/agi26-open-build-omegaclaw-fabricpc) contribution for FabricPC.

We started only about two days before the submission deadline, so this PR is intentionally unfinished. We will clean things up later and would very much appreciate **overall feedback** from the FabricPC maintainers on direction, API fit, and what belongs (or does not belong) in the examples tree.

### Planned cleanup (not done yet)
- Make the glucose dataset **downloadable** rather than shipping a large CSV in-repo
- **Exclude** reports, checkpoints, and other generated/run artifacts from the repo
- Further tidy docs, configs, and example surface area based on your feedback

## Why this PR

Most FabricPC examples today are toy / demo-scale. Our main contribution is a **real-domain use case**: blood-glucose forecasting from continuous glucose monitor (CGM) readings, so users can see FabricPC applied to something more interesting than MNIST-style demos.

## Science / what we did

We wanted to evaluate promising predictive-coding (PC) architectures on a domain we know well: **glucose prediction**.

Previously we developed a transformer-based **SugarOne** model that predicts future glucose values from past CGM device readings (wearable continuous glucose monitors), with additional covariates (e.g. carbs / insulin-related signals; related work also uses heart rate and steps). Because of the intersecting Open Build deadline, we only had ~two days, so we used a **simplified SugarOne-style architecture**: we cut out carbs, heart rate, and steps, and trained on a single-person (Livia) dataset to keep training tractable in the available time.

### Experiments
1. **FabricPC PC transformers** for glucose forecasting, with a **regular backprop** baseline of the same model for comparison. Early PC metrics were less competitive; we also hit **numeric stability** issues.
2. An **Optuna** hyperparameter-tuning pipeline to push PC further (search spaces, interruption-safe runs, report generators).
3. An exploratory **Hopfield / PC** path aimed at denoising glucose sequences. Exciting as an experiment, but it **did not improve** over the transformer / transformer-PC setups in our runs.

### Findings so far (on this dataset / setup)
- **PC underperformed** relative to the backprop-trained transformer baseline
- **Huber + PC** was **worse** than regular PC
- PC performed better on **shallow** networks with a **small number of heads and layers**

## What is in this draft
- `examples/glucose_*` — data loading, model, train/eval CLIs, Optuna tuning, Hopfield variant
- YAML search spaces under `examples/configs/`
- Tests for the glucose examples / tuning helpers
- Supporting docs and tuning progress reports (candidates to remove or slim down in cleanup)
- Bundled Livia CSV under `examples/data/` (planned to become a download)

## Ask for maintainers
High-level review welcome: Is a real glucose forecasting example a good fit for FabricPC? What should stay in-tree vs live as an external recipe? Any preferred patterns for datasets, tuning, and comparison to backprop that we should align to before we mark this ready for review?

Thanks — happy to iterate after the Open Build window.

Made with [Cursor](https://cursor.com)

### CODE REVIEW

**Comment by matthewbehrend** [comment]
# Review: PR #43 — Glucose forecasting example (draft)

Thanks for this, and apologies for the review latency. The work here is genuinely useful: you found a real library bug, built two library-level features that don't exist upstream, ran the first serious timeseries study on FabricPC, and documented the instability behavior in enough detail to be diagnostic. The answer to "how do we collaborate" is decomposition: this PR as it stands (59 files, +208k lines, now conflicted with 0.4.0) won't converge as a unit, but four smaller PRs extracted from it will. This review lays out that path, plus a reading of your results that I think changes the demo's framing.

All PRs should be cut fresh from upstream `main` — 0.4.0 landed since your branch point (`jax_setup.py` was replaced by `fabricpc/jax_config.py`, the Python floor moved to 3.11, and `optuna` moved to the `[experiments]` extra).

## 1. Bug-fix PR: StorkeyHopfield energy reduction — please submit now

Your change to `fabricpc/nodes/storkey_hopfield.py` is a genuine defect fix, not a workaround. The library contract is per-sample energy of shape `(batch,)` (`core/types.py`, `core/energy.py` docstrings; every state initializer allocates `energy=jnp.zeros((batch_size,))`, and energy is a `fori_loop` carry, so the shape is pinned). Every `EnergyFunctional` reduces over all non-batch axes with `tuple(range(1, ndim))`. `StorkeyHopfield` is the only node that adds an energy term outside the functional, and on `main` it reduces only `axis=-1`, so a rank>1 latent such as `(batch, seq, D)` produces a `(batch, seq)` term added to `(batch,)` energy: a shape error when `batch != seq`, and a silent mis-broadcast when they happen to be equal. Your glucose Hopfield model (`StorkeyHopfield(shape=(seq_len, embed_dim))`) is exactly the case `main` cannot run.

For the isolated PR:
- Drop the `if E_hopfield.ndim > 1` guard — `jnp.sum(x, axis=())` is already the identity on 1-D input, so the unconditional `jnp.sum(..., axis=tuple(range(1, E_hopfield.ndim)))` is simpler and matches `core/energy.py` exactly.
- Update the docstring: for rank>1 latents the term is now the sum over non-batch axes of `(s/2D) zᵀ(W² − W)z`.
- Add a direct unit test asserting `energy.shape == (batch,)` for a 3-D latent with `batch != seq`, mirroring the shape assertions in `tests/test_convolutional.py`. The existing `tests/test_storkey_hopfield.py` is 1-D-only, which is the coverage hole your example fell into.

## 2. Feature PR: `train_step_ipc` — invited, with small additions

This is a clean implementation of incremental PC (Salvatori et al., ICLR 2024) and nothing like it exists upstream. The energy aggregation matches `train_step` (same `in_degree > 0` filter, same `/ batch_size`), and the documented learning-rate-schedule interaction (the optimizer steps `infer_steps` times per batch, so schedule lengths must be scaled) is honored by your callers. Your `jnp.sum` inner reduction is also strictly better than the Python `sum` in `get_graph_param_gradient`; feel free to apply it there in the same PR.

Requested for the PR:
- At least one test (a convergence smoke test on a small graph is enough).
- Note in the docstring that it is reachable only from bespoke loops (`train_pcn` hardcodes `train_step`) and has no `pmap` counterpart yet. Both are fine for a first landing.
- Be aware of two PRs shipping around end of August: PR #36 (unified trainer) is changing the training-function surface; PR #47 (ePC) changes the inference solver API.

## 3. Feature PR: `InferenceAdam` — Park InferenceAdam until PR #47 lands and implement it against the new contract.

The Adam loop itself is sound JAX and fills a real gap (upstream has only `InferenceSGD` and `InferenceSGDNormClip`).
The blocker: the class overrides `run_inference`, but its `compute_new_latent` is a plain SGD step. `InferenceBase.inference_step` dispatches through `compute_new_latent`, and `inference_step` has callers other than `run_inference`: `run_inference_with_history` and `run_inference_with_full_history` in `fabricpc/utils/dashboarding/inference_tracking.py` (the Aim tracking path), and your own `train_step_ipc`. On all three paths `InferenceAdam` silently executes plain unclipped SGD — no momentum, no second moment, `max_norm` ignored.
In your branch `--infer-optimizer adam` and `--ipc` are independent flags in `examples/glucose_transformer.py`, and independent Optuna dimensions in the tuner. Every `(adam, ipc=True)` trial in your studies actually measured unclipped SGD, so any conclusions drawn about Adam under iPC need re-running.

The run_inference override hook exists for scan-based tracking and adaptive stopping. Stretching it to change the update rule resulted in duplicating the zero-grad and forward plumbing.
The right design is to put solver state in the step contract (the optax pattern) using a pytree threaded through the existing fori_loop/scan carries.

For example:
```python
class InferenceBase(ABC):                                                                                                                                                                                                                                                                                         
    @staticmethod                                                                                                                                                                                                                                                                                                 
    def init_solver_state(initial_state, clamps, structure, config):                                                                                                                                                                                                                                              
        return ()   # SGD siblings: empty pytree                                                                                                                                                                                                                                                                  
                                                                                                                                                                                                                                                                                                                  
    @staticmethod                                                                                                                                                                                                                                                                                                 
    def inference_step(params, state, solver_state, t, clamps, structure, config)
        -> (GraphState, solver_state)
```
Once the inference step carries state, the class family can be greatly simplified. One InferenceOptax(transform: optax.GradientTransformation, infer_steps) whose solver state is the OptState over the latent pytree replaces InferenceSGD, InferenceSGDNormClip, and InferenceAdam with optax.sgd, optax.adam, and optax.chain(clip, adam) compositions — optax is already a core dependency of the training loop.

Required fix:
- Design the stateful step contract with Adam as its first client.
- Restore `.conj()` in the norm computation for parity with `InferenceSGDNormClip` (identical for real arrays, correct for complex).
- Tests and one clean benchmark with the fix in place showing where Adam inference actually helps.
- It is also reasonable to park this until ePC ships, since ePC targets the same instability with a stronger mechanism.


## 4. Demo PR: glucose timeseries forecasting — wanted, with a specific shape

A timeseries example would be a novel addition to `examples/` and we'd like to have it. The conventions it must match (see `transformer_demo.py`, `PC_backprop_compare.py`): one self-contained file, module docstring with an architecture sketch and a `Usage:` block, argparse, `create_*_model()` factories, `--mode {pc,backprop}` in the same file, plain-text table output. No YAML configs, no typer/dotenv, no plotting, no bundled data, and no changes outside the demo file and its tests.

Concretely:

1. **Dataset on Hugging Face, not in the repo.** Your `glucose_data.py` already sketches an HF fallback download; make that the only path. Publish the CSV with the `Recommended Split` column filled — in the bundled file it is empty on all 139,613 rows, which is why every reported run fell back to the random-by-sequence split.
2. **Chronological split only.** With the random split, test sequences can precede training sequences in time. Document the remaining caveats in the docstring: single subject, stride-1 overlapping windows, so results measure within-subject temporal generalization.
3. **A persistence baseline** (last-value carry-forward, i.e. predict the final observed glucose for the whole horizon) printed alongside both models. There is currently no naive baseline anywhere in the branch, and for 60-minute CGM forecasting persistence plausibly lands in the same 20–30 mg/dL MAE band as every model reported. Without it, none of the numbers is interpretable.
4. **Matched arms.** PC and backprop on the same graph, same seed, same optimizer settings, at the tuned shallow geometry (context 64, depth 1), with weight-gradient clipping at global norm 1.0 and `InferenceSGDNormClip` on the latents — the configuration your own log shows is stable for both modes.
5. **Keep the good parts, drop the bookkeeping.** The load→split→window→train-only-min-max pipeline in `glucose_data.py` and the three custom nodes in `glucose_model.py` (continuous embedding, multi-scale MHA residual, regression readout) collapse nicely into one demo file. The resume/checkpoint/`config.json`/`comparison.json` machinery in `glucose_transformer.py` (roughly two-thirds of the file) stays in your fork.
6. **No dependency changes.** `typer`, `python-dotenv`, and `pyyaml` as core dependencies won't be accepted; avoid pandas or put it behind an extra. Console-script entries under `[project.scripts]` are a convention of your fork, not upstream — propose that separately if you want it.

## 6. Stability, depth, and ePC timing

Your instability findings mare consistent with our own experience. Gradient clipping, shallow networks, and low-rate inference are the available mitigations today, but they contain the instability rather than remove it.

The structural remedy is ePC, which is shipping this month (design in draft PR #47): it targets stability directly, trains deep-layer weights that state-based PC leaves essentially untrained, and converges in roughly 10 inference steps instead of ~120. My recommendation is to land the demo now at the shallow stable configuration and revisit the model development — depth, inference budget, and the PC-vs-backprop gap — once ePC is in.

## 7. Github Issue #46 and the data path

Issue #46 specifies a Grain-based random-access data path; its source protocol (`__len__`/`__getitem__`) is exactly what a Hugging Face `datasets.Dataset` satisfies, so a glucose/HF source is a concrete conforming-source test case for that spec. The demo does not need to wait for it — ship with the loader embedded in the demo file, which explicitly joins #46's migration inventory (as `storkey_hopfield_recall.py` already does) — but if you want to contribute there on the way, the work is open and unassigned.

## 8. Small defects in the branch, for your own housekeeping

- `tests/test_glucose_hopfield_tuning.py:34` asserts `TARGET_OPTUNA_MAE == 19.876` but `examples/glucose_hopfield_tuning.py:71` defines `19.408`; the test fails as committed.
- The master report's "How to run" table references `scripts/summarize_glucose_tuning.py`, `scripts/run_hopfield_optuna_wsl.sh`, and `glucose-transformer-tune-update-budget`, none of which exist on the branch.
- The Hopfield v1 study's "best" (20.30) comes from a PRUNED trial in a study with zero completed trials, but is listed in the cross-study table as comparable.
- `hopfield_strength` is sampled for the `baseline` variant, which has no Hopfield node, so the top rows of the WSL studies are the same architecture differing only by a no-op parameter and seed.

## Summary

Please open, in order: (1) the StorkeyHopfield energy fix, (2) `train_step_ipc`, (3) `InferenceAdam` once the `inference_step` contract is fixed and benchmarked (or park it until ePC), and (4) the self-contained glucose demo with the HF-hosted dataset, chronological split, persistence baseline, and matched arms. Everything else — reports, report generators, the Optuna harnesses, YAML spaces, the bundled CSV, and the Hopfield variants — stays in your fork, and the Hopfield negative result deserves a write-up. I'll review each of those promptly; sorry again that this one sat.
